test/test_stpncpy_chk.c

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2024 Frederick H. G. Wright II <fw@fwright.net>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+/*
+ * This provides test cases for the stpncpy() security wrapper.
+ * There may or may not be compile-time warnings in some cases, but
+ * the final pass/fail should be as expected.
+ *
+ * Informational messages regarding the relevant compile flags are also
+ * included.
+ */
+
+/* Allow deferring the stdio.h include */
+int printf(const char *format, ...);
+
+#define PRINT_VAR(x) printf("    %s = %lld\n", #x, (long long) x)
+#define PRINT_UNDEF(x) printf("    " #x " is undefined\n")
+
+void
+print_before(void)
+{
+  printf("  Before string.h include:\n");
+  #ifdef _FORTIFY_SOURCE
+  PRINT_VAR(_FORTIFY_SOURCE);
+  #else
+  PRINT_UNDEF(_FORTIFY_SOURCE);
+  #endif
+  #ifdef _USE_FORTIFY_LEVEL
+  PRINT_VAR(_USE_FORTIFY_LEVEL);
+  #else
+  PRINT_UNDEF(_USE_FORTIFY_LEVEL);
+  #endif
+}
+
+#include <string.h>
+
+void
+print_after(void)
+{
+  printf("  After  string.h include:\n");
+  #ifdef _FORTIFY_SOURCE
+  PRINT_VAR(_FORTIFY_SOURCE);
+  #else
+  PRINT_UNDEF(_FORTIFY_SOURCE);
+  #endif
+  #ifdef _USE_FORTIFY_LEVEL
+  PRINT_VAR(_USE_FORTIFY_LEVEL);
+  #else
+  PRINT_UNDEF(_USE_FORTIFY_LEVEL);
+  #endif
+}
+
+#if defined(_USE_FORTIFY_LEVEL) && _USE_FORTIFY_LEVEL > 1
```

Comment:
In `string.h`, checks appear to be enabled when `_USE_FORTIFY_LEVEL` > 0.

But here, we're only enabling tests when the value is greater than 1.

Admittedly my knowledge of these preproc symbols is lacking. (Particularly in terms of the semantics related to their values.)

But shouldn't tests be enabled, if the checks are...?

Reply:
You're right, though it's not clear when it makes a difference.  I had '2' on the brain because that's the only value that ever gets set by default.  I'm not sure what the difference is, if any, between '1' and '2'.  But for consistency, I changed the threshold in the test and added another "forced" test setting it to '1' for good measure.  And retested on all the above systems.

I didn't retest again after fixing the typo in the Makefile comment. :-)

Reply:
> But for consistency, I changed the threshold in the test and added another "forced" test setting it to '1' for good measure.  And retested on all the above systems.

Sounds good!